File: pong_server/src/game_server/Game/PlayerLocation.py
from Game.MatchLocation import MatchLocation


class PlayerLocation(object):
    def __init__(self,
                 game_round: int,
                 match: int,
                 player_index: int,
                 is_looser: bool = False):
        self.is_looser: bool = is_looser

        if is_looser:
            self.match_location: MatchLocation = MatchLocation(-1, -1)
        else:
            self.match_location: MatchLocation = MatchLocation(game_round, match)

        self.player_index: int = player_index

    def to_json(self) -> dict:
        return {
            'is_looser': self.is_looser,
            'match_location': self.match_location.to_json(),
            'player_index': self.player_index
        }

File: front/app/src/js/components/game/Scene/PlayerLocation.js
import {Scene} from './Scene';

export class PlayerLocation {
  #isLooser;
  #matchKey;
  #playerIndex;

  constructor(playerLocationJson) {
    this.#isLooser = playerLocationJson['is_looser'];
    this.#matchKey = Scene.convertMatchLocationToKey(
        playerLocationJson['match_location'],
    );
    this.#playerIndex = playerLocationJson['player_index'];
  }

  getPlayerMatchFromScene(scene) {
    if (this.#isLooser === true) {
      return null;
    }
    return scene.getMatchFromKey(this.#matchKey);
  }

  getPlayerFromScene(scene) {
    if (this.#isLooser === true) {
      return scene.loosers[this.#playerIndex];
    }
    return this.getPlayerMatchFromScene(scene).players[this.#playerIndex];
  }

  get playerIndex() {
    return this.#playerIndex;
  }

  get isLooser() {
    return this.#isLooser;
  }
}
